### PR TITLE
torchmultimodal CLIP

### PR DESCRIPTION
**Pull request by msaroufim**
co-authored with @ankitade

### CODE REVIEW

File: torchbenchmark/models/clip/requirements.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+git+https://github.com/facebookresearch/multimodal.git@bf42547c5fcfd063eba4f036def0c57309c2ee57
```
Comment (ankitade) [inline torchbenchmark/models/clip/requirements.txt:1]: why not just use the nightly?

Comment (msaroufim) [reply torchbenchmark/models/clip/requirements.txt:1]: It's what was in your notebook, I just copied that - can use nightly if you prefer and if no breaking changes are expected

Comment (msaroufim) [reply torchbenchmark/models/clip/requirements.txt:1]: i fixed

**Comment by msaroufim** [comment]
@ankitade could you please give this a second take?

File: torchbenchmark/models/clip/requirements.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+git+https://github.com/facebookresearch/multimodal.git
```
Comment (ankitade) [inline torchbenchmark/models/clip/requirements.txt:1]: nit: just torchmultimodal should work

File: torchbenchmark/models/clip/install.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import os
+import subprocess
+import sys
+
+def pip_install_requirements():
+    subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '-r', 'requirements.txt'])
+
+def download_data():
+    subprocess.check_call(['wget', '-O', './data', 'https://upload.wikimedia.org/wikipedia/commons/thumb/9/91/Pizza-3007395.jpg/2880px-Pizza-3007395.jpg'])
```
Comment (ankitade) [inline torchbenchmark/models/clip/install.py:9]: any legal requriements on the image? i was just using a random one in my notebook

File: torchbenchmark/models/clip/__init__.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+from torchmultimodal.modules.losses.contrastive_loss_with_temperature import (
+    ContrastiveLossWithTemperature,
+)
+from PIL import Image
+import math
+
+
+
+class Model(BenchmarkModel):
+    DEFAULT_EVAL_BSIZE = 32
```
Comment (ankitade) [inline torchbenchmark/models/clip/__init__.py:21]: not used?

File: torchbenchmark/models/clip/__init__.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+        # Zero the gradients
+        optimizer.zero_grad()
+
+        # Forward pass
+        image_embedding, text_embedding = self.model(self.image_tensor, self.text_tensor)
+        score = image_embedding @ text_embedding.t()
+
+        # Compute loss
+        loss = loss_fn(score)
```
Comment (ankitade) [inline torchbenchmark/models/clip/__init__.py:64]: shouldnt the loss take in image and text embeddings and not the score?

File: torchbenchmark/models/clip/__init__.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        self.model.train()
+
+        # Create optimizer
+        loss_fn = ContrastiveLossWithTemperature()
+        optimizer = torch.optim.AdamW(
+            list(self.model.parameters()) + list(loss_fn.parameters()),
+            lr=5.0e-4,
+            weight_decay=1.0e-4,
+            eps=1.0e-6,
+            foreach=True,
```
Comment (ankitade) [inline torchbenchmark/models/clip/__init__.py:53]: do you care about making these configurable? like foreach was somehow increasing step time for me.

File: torchbenchmark/models/clip/install.py (hunk 3 above)
Comment (msaroufim) [reply torchbenchmark/models/clip/install.py:9]: it's a cc0 image from the wikipedia page for pizza lol https://en.wikipedia.org/wiki/Pizza#/media/File:Pizza-3007395.jpg so should be good

File: torchbenchmark/models/clip/__init__.py (hunk 6 above)
Comment (msaroufim) [reply torchbenchmark/models/clip/__init__.py:53]: Not particularly I can just remove foreach

File: torchbenchmark/models/clip/__init__.py (hunk 4 above)
Comment (msaroufim) [reply torchbenchmark/models/clip/__init__.py:21]: so question here is clip typically used in bs=1 regimes? We were surprised to see that was the case for sam for example

cc @cpuhrsch

File: torchbenchmark/models/clip/__init__.py (hunk 5 above)
Comment (msaroufim) [reply torchbenchmark/models/clip/__init__.py:64]: todo for self: this whole test is getting skipped, not sure why yet

Comment (msaroufim) [reply torchbenchmark/models/clip/__init__.py]: hmm i tried a few variants of `metadata.yaml` - @xuzhao9 do you see any obvious issues?

Comment (msaroufim) [reply torchbenchmark/models/clip/__init__.py]: AH found the issue, needed to add `    DEFAULT_TRAIN_BSIZE = 1` to `__init__.py`

File: torchbenchmark/models/clip/__init__.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        )
+
+        total_loss = 0
+        # Iterate over each text description
+        for i in range(len(self.text)):
+            # Zero the gradients
+            optimizer.zero_grad()
+
+            # Forward pass
+            image_embedding, text_embedding = self.model(self.image_tensor, self.text_tensor[i].unsqueeze(0))
```
Comment (ankitade) [inline torchbenchmark/models/clip/__init__.py:61]: ideally you want a batch of pairs vs doing this. but prob doesn't matter for benchmarking ?

**Review by ankitade — APPROVED** [review]
lgtm

File: torchbenchmark/models/clip/__init__.py (hunk 7 above)
Comment (msaroufim) [reply torchbenchmark/models/clip/__init__.py:61]: actually i would prefer the batched implementation!

File: torchbenchmark/models/clip/__init__.py (hunk 4 above)
Comment (msaroufim) [reply torchbenchmark/models/clip/__init__.py:19]: Added support for batch size

File: torchbenchmark/models/clip/__init__.py (hunk 7 above)
Comment (msaroufim) [reply torchbenchmark/models/clip/__init__.py]: fixed!

File: torchbenchmark/models/clip/__init__.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    def get_module(self):
+        return self.model, (self.image_tensor, self.text_tensor)
+
+            
+    def train(self):
+        self.model.train()
+
+        # Create optimizer
+        loss_fn = ContrastiveLossWithTemperature()
+        optimizer = torch.optim.AdamW(
```
Comment (xuzhao9) [inline torchbenchmark/models/clip/__init__.py:51]: It would be better to move the initialization of loss_fn and optimizer in `__init__()`

Comment (msaroufim) [reply torchbenchmark/models/clip/__init__.py]: done!

**Review by xuzhao9 — APPROVED** [review]
LGTM. We may remove the explicit jit option soon: https://github.com/pytorch/benchmark/pull/1787

File: torchbenchmark/models/clip/requirements.txt (hunk 2 above)
Comment (xuzhao9) [inline torchbenchmark/models/clip/requirements.txt:1]: We suggest to pin the version of this dependency with commit hash, because when we are benchmarking pytorch nightly, and changes from upstream model code might cause false positive performance signals.

File: torchbenchmark/models/clip/metadata.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+devices:
+  NVIDIA A100-SXM4-40GB:
+    eval_batch_size: 1
+    eval_train_size: 1
```
Comment (xuzhao9) [inline torchbenchmark/models/clip/metadata.yaml:4]: Sorry what is `eval_train_size`?

Comment (msaroufim) [reply torchbenchmark/models/clip/metadata.yaml:4]: typo

**Comment by msaroufim** [comment]
Hmm the phabricator UI wasn't showing a land happening for me, I'm confused why this landed, will monitor CI and if it fails I'll forward fix - sorry!
